internal/controllers/machinedeployment/machinedeployment_status.go

```diff
@@ -0,0 +1,502 @@
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:    clusterv1.MachineDeploymentScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineDeploymentScalingDownInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should never happen).
+	if machineDeployment.Spec.Replicas == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentScalingDownV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineDeploymentScalingDownWaitingForReplicasSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *machineDeployment.Spec.Replicas
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+	currentReplicas := mdutil.GetReplicaCountForMachineSets(machineSets)
+
+	// Scaling down.
+	if currentReplicas > desiredReplicas {
+		message := fmt.Sprintf("Scaling down from %d to %d replicas", currentReplicas, desiredReplicas)
+		staleMessage := aggregateStaleMachines(machines)
+		if staleMessage != "" {
+			message += fmt.Sprintf(" and %s", staleMessage)
+		}
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:    clusterv1.MachineDeploymentScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionTrue,
+			Reason:  clusterv1.MachineDeploymentScalingDownV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	// Not scaling down.
+	v1beta2conditions.Set(machineDeployment, metav1.Condition{
+		Type:   clusterv1.MachineDeploymentScalingDownV1Beta2Condition,
+		Status: metav1.ConditionFalse,
+		Reason: clusterv1.MachineDeploymentNotScalingDownV1Beta2Reason,
+	})
+}
+
+func setMachinesReadyCondition(ctx context.Context, machineDeployment *clusterv1.MachineDeployment, machines collections.Machines) {
+	log := ctrl.LoggerFrom(ctx)
+	// If we got unexpected errors in listing the machines (this should never happen), surface them.
+	if machines == nil {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:    clusterv1.MachineDeploymentMachinesReadyV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineDeploymentMachinesReadyInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	if len(machines) == 0 {
+		v1beta2conditions.Set(machineDeployment, metav1.Condition{
+			Type:   clusterv1.MachineDeploymentMachinesReadyV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: clusterv1.MachineDeploymentMachinesReadyNoReplicasV1Beta2Reason,
+		})
+		return
+	}
+
+	readyCondition, err := v1beta2conditions.NewAggregateCondition(
+		machines.UnsortedList(), clusterv1.MachineReadyV1Beta2Condition,
+		v1beta2conditions.TargetConditionType(clusterv1.MachineDeploymentMachinesReadyV1Beta2Condition),
+	)
+	if err != nil {
+		log.Error(err, "Failed to aggregate Machine's Ready conditions")
```

Comment:
Random idea. WDYT about scanning controller logs after e2e tests for errors that should never occur and fail the e2e test jobs if we find any? (just like we check for no panics)

Would be a follow-up task of course

Reply:
We can discuss as a follow up, but we need to find a definition of "errors that should never occur" that is easy to maintain

Reply:
Yup, definitely

Reply:
Added a note on my list, will create an issue for it